listenbrainz_spark/stats/user/listening_activity.py

```diff
@@ -0,0 +1,192 @@
+from datetime import datetime
+from typing import Iterator
+
+from flask import current_app
+
+import listenbrainz_spark
+from data.model.user_listening_activity import UserListeningActivityStatMessage
+from listenbrainz_spark.constants import LAST_FM_FOUNDING_YEAR
+from listenbrainz_spark.path import LISTENBRAINZ_DATA_DIRECTORY
+from listenbrainz_spark.stats import (adjust_days, adjust_months, get_day_end,
+                                      get_month_end, get_year_end,
+                                      replace_days, replace_months, run_query)
+from listenbrainz_spark.stats.user.utils import (filter_listens,
+                                                 get_last_monday,
+                                                 get_latest_listen_ts)
+from listenbrainz_spark.utils import get_listens
+from pyspark.sql.functions import collect_list, sort_array, struct
+from pyspark.sql.types import (StringType, StructField, StructType,
+                               TimestampType)
+
+time_range_schema = StructType((StructField('time_range', StringType()), StructField(
+    'start', TimestampType()), StructField('end', TimestampType())))
+
+
+def get_listening_activity():
+    """ Calculate number of listens for each user in time ranges given in the 'time_range' table """
+    int_result = run_query("""
+            SELECT listens.user_name
+                 , time_range.time_range
+                 , count(listens.user_name) as listen_count
+              FROM listens
+              JOIN time_range
+                ON listens.listened_at >= time_range.start
+               AND listens.listened_at <= time_range.end
+          GROUP BY listens.user_name
+                 , time_range.time_range
+            """)
+    int_result.createOrReplaceTempView('int_result')
+
+    result = run_query("""
+            SELECT dist_user_name.user_name
+                 , time_range.time_range
+                 , to_unix_timestamp(time_range.start) as from_ts
+                 , to_unix_timestamp(time_range.end) as to_ts
+                 , ifnull(int_result.listen_count, 0) as listen_count
+              FROM (SELECT DISTINCT user_name FROM listens) dist_user_name
+        CROSS JOIN time_range
+         LEFT JOIN int_result
+                ON int_result.user_name = dist_user_name.user_name
+               AND int_result.time_range = time_range.time_range
+            """)
+
+    iterator = result \
+        .withColumn("listening_activity", struct("from_ts", "to_ts", "listen_count", "time_range")) \
+        .groupBy("user_name") \
+        .agg(sort_array(collect_list("listening_activity")).alias("listening_activity")) \
+        .toLocalIterator()
+
+    return iterator
+
+
+def get_listening_activity_week() -> Iterator[UserListeningActivityStatMessage]:
```

Comment:
a lot of the code in this and the corresponding month and year function is duplicated, can we deduplicate it somehow?